Mark chat resolved (Issue #27) and get next student (Issue #66)

rorochaudhary opened the pull request:
This PR closes #27 and closes #66.

Changes/Features:
- `StudentStatus` enum representing the possible states a student can take in the office hours workflow ('WAITING', 'CONVERSING', 'RESOLVED') added to `bot/utilities/Global.ts`. This is reflected in the QueueEntry structure and the database.
- The database schema for `QueueEntry.resolved` field is changed from a `boolean` to `nvarchar`. Newly inserted entries contain a default `resolved` value of 'WAITING'. 
- `/bot/api/updateQueueEntryResolved.ts` contains the ORM mapping to update the `resolved` field.
- `mark student complete` and `get next student` bot commands that update the `resolved` fields appropriately. Enforces the idea that only one student can be in a `CONVERSING` state at a time and must be marked as `RESOLVED` before an instructor can move on and help the next student.
- supporting Queue and QueueEntry methods to accomplish the above.

Screenshots of functionality:
![image](https://user-images.githubusercontent.com/40576340/155056662-63136f92-9f75-4068-94d1-9ef062407d52.png)
![image](https://user-images.githubusercontent.com/40576340/155056719-b5198ce6-5abd-4961-91fc-88613b07651a.png)

rorochaudhary committed:
QueueEntry and QueueEntity resolved field changed from bool to enum, starting changes for mark student complete, untested change

rorochaudhary committed:
mark student complete command started, untested, setResolvedState added to QueueEntry

rorochaudhary committed:
add isEmpty and isNotEmpty to Queue structure, update mark student complete failure msg

rorochaudhary committed:
added 'get next student' command, added findfirstWaiting to Queue structure, updates persist to backend

rorochaudhary committed:
code formatting with prettier

Jaqito commented:
Took a look at this. Did you notice that the in the screenshots in the queue UserId is undefined? I'm not sure what changes were made to the DB that caused this but it's making the bot not work. Right now It's trying to find a student ID where it == undefined but that is every student. Think this bug will need to be fixed before It can be merged.

rorochaudhary commented:
> Took a look at this. Did you notice that the in the screenshots in the queue UserId is undefined? I'm not sure what changes were made to the DB that caused this but it's making the bot not work. Right now It's trying to find a student ID where it == undefined but that is every student. Think this bug will need to be fixed before It can be merged.

Looked into why that is the case, thanks for catching that! The results are persisted to the DB fine, the user ID exists there. The code currently takes the results of the SQL insert and then generates an entry into our queue data structure that is used by the bot. Turns out the returned results of the SQL insert don't return the user ID, so it doesn't end up in our in-app queue. The DB action itself is unaffected.

Jaqito commented:
Coding here from my GitHub app on my phone so sorry lol.

instead of not returning the columns on the addQueueEntry call we can update it so that it's more similar to https://github.com/beaver-hours/BeaverHours/blob/master/bot/api/addQueueToDb.ts

You can see how we explicitly return the column names we want back when we addQueueToDb but don't do so on addQueueEntryToDb

@rorochaudhary

rorochaudhary commented:
> Coding here from my GitHub app on my phone so sorry lol.
> 
> instead of not returning the columns on the addQueueEntry call we can update it so that it's more similar to https://github.com/beaver-hours/BeaverHours/blob/master/bot/api/addQueueToDb.ts
> 
> You can see how we explicitly return the column names we want back when we addQueueToDb but don't do so on addQueueEntryToDb
> 
> @rorochaudhary

Thanks for that reference. Between this and Richard having a fix also in the pipeline in another PR this bug looks to be handled.

richardcoucoules commented on bot/utilities/Queue.ts:
```diff
@@ -36,6 +37,14 @@ export default class Queue {
+  isEmpty(): boolean {
+    return this.length == 0;
+  }
+
```
It seems a little curious to have both an `isEmpty` method and an `isNotEmpty` method; `isNotEmpty` returns the negation of `isEmpty`, so the user can just use the value `!Queue.isEmpty`.

rorochaudhary replied:
Good point. I debated the same thing and chose to err on the side of explicitness in this case. It would make sense for an interface being more widely interacted with/reused. Removed!

rorochaudhary committed:
fix get next student error for non-empty queue with not waiting students, output grammar fix

rorochaudhary committed:
removing unnecessary Queue.isNotEmpty method

rorochaudhary committed:
code formatting with prettier

rorochaudhary committed:
fix rebase artifact, add missing break statement (fixes code in master)

rorochaudhary commented:
Rebased onto new changes in master and added a missing `break` statement within the `view active queue` case, resolving a bug that is currently within the master branch.

richardcoucoules reviewed (approved):
Great work!
